### DISCUSSION TITLE
support OverloadResolutionPriority for extension methods

### DISCUSSION THREAD

**Discussion by BarionLP** [dotnet/runtime#108963] [Ideas] [answered]
I recently had the issue of ambiguous calls with extension methods from different classes (#108960)
when trying to fix this with the `OverloadResolutionPriorityAttribute` it unexpectedly did not work because the attribute only seems to affect methods in the same class.
I think this attribute should affect overloading in general and not only in specific scenarios.

**Comment by huoyaoyuan** [comment, marked as answer]
This is the wrong repository for this topic. It's a compile time feature that belongs to csharplang repository.

This was explicitly rejected in the early design of the feature: https://github.com/dotnet/csharplang/blob/main/meetings%2F2024%2FLDM-2024-05-13.md

**Reply by BarionLP** [reply]
> This is the wrong repository for this topic. It's a compile time feature that belongs to csharplang repository.

sorry

> This was explicitly rejected in the early design of the feature:

after reading the reasons I do understand the decision but I do want to raise attention to my problem. maybe a different solution can be found?

**Reply by julealgon** [reply]
> maybe a different solution can be found?

@BarionLP have you considered creating a new extension methods class with 2 methods, each redirecting to your specific extensions, and then annotating _those_ with the overload attribute?

Sure, it is not even close to ideal, but might serve as a workaround for you.

**Reply by BarionLP** [reply]
> > maybe a different solution can be found?
> 
> @BarionLP have you considered creating a new extension methods class with 2 methods, each redirecting to your specific extensions, and then annotating _those_ with the overload attribute?
> 
> Sure, it is not even close to ideal, but might serve as a workaround for you.

might work in other situations but in this case the signature is too similar (only differs by generic constrains)

**Reply by julealgon** [reply]
> might work in other situations but in this case the signature is too similar (only differs by generic constrains)

You might _still_ make it work if you "force" the signature to be different by introducing a "dummy" default parameter to one of them @BarionLP .

For example, there will cause problems since you can't have overloads differ only by generic constraints:
```csharp
public static class MyExtensions
{
    public static void Do<T>(T thing) => ExtensionClass1.Do(thing);

    [OverloadResolutionPriority(1)]
    public static void Do<TSomething>(TSomething thing)
        where TSomething : ISomething
        => ExtensionsClass2.Do(thing);        
}
```

But _this_ should work:
```csharp
public static class MyExtensions
{
    public static void Do<T>(T thing) => ExtensionClass1.Do(thing);

    [OverloadResolutionPriority(1)]
    public static void Do<TSomething>(TSomething thing, object dummy = null)
        where TSomething : ISomething
        => ExtensionsClass2.Do(thing);
}
```

**Reply by BarionLP** [reply]
> But _this_ should work:

thank you @julealgon, but unfortunally I can't get it to work with this
this is the issue I'm trying to solve https://github.com/thomhurst/TUnit/issues/922
file: https://github.com/thomhurst/TUnit/blob/main/TUnit.Assertions/Assertions/Generics/GenericIsExtensions_Namespaced.cs#L14
